### PR TITLE
Fire and forget

### PR DESCRIPTION
**Pull request by delvedor**
I think this could be nice to have, is just a wrapper over `request` to make the api nicer.
Let me know what do you think!

```js
instance.fire({
  key: 'some data',
  hello: 42
})
```

### CODE REVIEW

**Comment by delvedor** [comment]
Yes we could, in fact this pr is intended to have a nicer api, read alias, to do this operation.
Furthermore we provide a debug log and by wrapping the request in a nextTick the execution is not blocking, as you can see [here](https://github.com/delvedor/upring/blob/c4c9988c2967e518e782ecff5ce917783b54c937/test/fire-and-forget.test.js#L47).
What I'm wondering is what gonna happen if the function that handles that request does not call `reply`?

**Comment by mcollina** [comment]
Why do you want it to be deferred?

Plus, if reply() is not called... then nothing will happen and part of the request will hang.

I think fire() should send a new type of message that does not need a reply. I think we might need some support from tentacoli as well to implement it properly.

**Comment by delvedor** [comment]
> Why do you want it to be deferred?

Because as user I expect that is "fire and forget" and not "fire and wait a little bit and then forget".
For example, [here](https://github.com/delvedor/upring/blob/c4c9988c2967e518e782ecff5ce917783b54c937/test/fire-and-forget.test.js#L47) without the `nextTick` the test will fail.

> Plus, if reply() is not called... then nothing will happen and part of the request will hang.

I'd imagined.

> I think fire() should send a new type of message that does not need a reply. I think we might need some support from tentacoli as well to implement it properly.

I like this a lot, I'll think about a solution there :)

File: upring.js [hunk 1]
```diff
@@ -257,10 +257,59 @@ UpRing.prototype.request = function (obj, callback, _count) {
...
+  if (!this.isReady) {
+    this.once('up', this.fire.bind(this, obj, callback))
+    return
+  }
+
+  if (this._hashring.allocatedToMe(obj.key)) {
+    this.logger.trace({ msg: obj }, 'local call')
+    this._dispatch(obj, dezalgo(callback.bind(this)))
+  } else {
+    let peer = this._hashring.lookup(obj.key)
```
Comment (mcollina) [inline upring.js]: why let?

File: upring.js [hunk 2]
```diff
@@ -257,8 +257,69 @@ UpRing.prototype.request = function (obj, callback, _count) {
...
+    this.logger.trace({ msg: obj, peer }, 'remote call')
+
+    let upring = peer.meta.upring
```
Comment (mcollina) [inline upring.js]: const?

File: upring.js [hunk 3]
```diff
@@ -257,8 +257,69 @@ UpRing.prototype.request = function (obj, callback, _count) {
...
+    conn.fire(obj, (err) => {
+      if (err) {
+        // the peer has changed
+        if (this._hashring.lookup(obj.key).id !== peer.id || conn.destroyed) {
+          // TODO make this dependent on the gossip interval
+          setTimeout(retry, 500, this, 'fire', obj, callback, _count)
+          return
+        }
+      }
+      callback.call(this, err)
```
Comment (mcollina) [inline upring.js]: why callback.call?

File: upring.js (hunk 1 above)
Comment (delvedor) [inline upring.js]: Don't know, you tell me :D
This is a copy and past of the [request](https://github.com/upringjs/upring/blob/master/upring.js#L204-L256) function!

File: upring.js (hunk 3 above)
Comment (delvedor) [inline upring.js]: To support [this](https://github.com/delvedor/upring/blob/d30b1569bb852477ff678dfebc7f569ecf183247/upring.js#L315-L319).

File: upring.js (hunk 1 above)
Comment (mcollina) [inline upring.js]: interesting one :D

File: upring.js (hunk 3 above)
Comment (mcollina) [inline upring.js]: why don't you do `fireCallback.bind(this)` and assign it to the object, so you can reuse it?

File: upring.js (hunk 1 above)
Comment (delvedor) [inline upring.js]: I can update them if you want :)

Comment (mcollina) [inline upring.js]: ahhaha yes thanks!

File: upring.js [hunk 4]
```diff
@@ -257,8 +258,69 @@ UpRing.prototype.request = function (obj, callback, _count) {
...
+UpRing.prototype.fire = function (obj, callback, _count) {
+  callback = callback || this._fireCallback
+  if (!this.isReady) {
+    this.once('up', this.fire.bind(this, obj, callback))
+    return
+  }
+
+  if (this._hashring.allocatedToMe(obj.key)) {
+    this.logger.trace({ msg: obj }, 'local call')
+    this._dispatch(obj, dezalgo(callback))
```
Comment (mcollina) [inline upring.js]: I think this should not call the callback. The callback should be called anyhow, as it has no feedback, but just error on delivery. In this case, the delivery is guaranteed.

Comment (delvedor) [inline upring.js]: You mean something like this?
```js
this.logger.trace({ msg: obj }, 'local call')
callback()
this._dispatch(obj, noop)
```

Comment (mcollina) [inline upring.js]: yes

**Comment by mcollina** [comment]
Thx!

**Pull request merged**